### ISSUE TITLE
Migrations Issue In Azure With Orchard_Framework_DataMigrationRecord Table

### ISSUE THREAD

**Issue by bleroy** [bleroy/OrchardImportTest#1950]
When upgrading modules in Orchard while the site is running in Azure causes
the migrations to run, but the record in the
Orchard_Framework_DataMigrationRecord table only gets incremented by a value
of 1 no matter how many updates run.  
  
Steps to reproduce this error are as follows...  
  
Create a module that requires a PartRecord and has a Migrations.cs with only a
Create() method in it.  
Deploy code to Azure and enable the module.  
Update module and create two update methods. UpdateFrom1() and UpdateFrom2().  
Deploy code to Azure and see get the message that you need to update the
module.  
Run the update and check the Orchard_Framework_DataMigrationRecord record for
the module.  
  
This issue will be that is has only incremented the version to 2 when it
should be 3.  
  
This works in our local Dev environment without issue, but in Azure it fails
to update it correct.  

Work Item Details
--------------------
**Original CodePlex Issue:**	[Issue 18060](http://orchard.codeplex.com/workitem/18060)
**Status:**	Closed
**Reason Closed:**	Unassigned
**Assigned to:**	Unassigned
**Reported on:**	Aug 3, 2011 at 7:48 PM
**Reported by:**	vrossi
**Updated on:**	Mar 25, 2014 at 11:07 PM
**Updated by:**	sebastienros
**Closed on:**	Mar 25, 2014 at 11:07 PM
**Closed by:**	sebastienros
**Status:**	Closed
**Reason Closed:**	Unassigned
**Assigned to:**	Unassigned
**Reported on:**	Aug 3, 2011 at 7:48 PM
**Reported by:**	vrossi
**Updated on:**	Mar 25, 2014 at 11:07 PM
**Updated by:**	sebastienros
**Closed on:**	Mar 25, 2014 at 11:07 PM
**Closed by:**	sebastienros

**Comment by bleroy** [comment]
On *2011-08-04 04:31:25 UTC*, **@BertrandLeRoy** commented:

We should probably block module updates in Azure. In the current state of things, that can't end well. Modules should be updated on a dev machine and then the package should be re-uploaded.

**Comment by bleroy** [comment]
On *2011-08-04 22:29:38 UTC*, **@AustinBergstrom** commented:

What is it about the current state of things that would cause module updates to not end well in Azure? It seems silly to not support such a key mechanism for developing these modules in iterations.

If it comes to be that Orchard will simply not support module updates in Azure and we are required to upload a new package each time the module is updated, will this cause problems with existing database records of said modules?

**Comment by bleroy** [comment]
On *2011-08-05 01:21:19 UTC*, **@BertrandLeRoy** commented:

Azure is an architecture that is built to scale by adding instances on the fly. The file system is not synchronized between these instances (a new instance just re-deploys the initial package), which means that you can't reliably change files on the server.
 Modules are a bunch of source code files, at least for now. If you wanted to support module modifications on Azure, you would need to store the code for those modules in blob storage or the database. As you can imagine this is not trivial.
Database data and metadata is not affected by this.

**Comment by bleroy** [comment]
On *2011-08-05 03:13:16 UTC*, **@BertrandLeRoy** commented:

Migrations should work.

**Comment by bleroy** [comment]
On *2011-08-05 03:14:52 UTC*, **@BertrandLeRoy** commented:

Just to be sure: when you say "deploy", you mean recreating the Azure package and redeploy that?

**Comment by bleroy** [comment]
On *2011-08-05 03:46:46 UTC*, **@vrossi** commented:

Correct. We are deploying an Azure package. The migrations work (meaning the tables and columns are created correctly) but the version number is incorrect, as stated in my original example.

**Comment by bleroy** [comment]
On *2011-08-05 04:58:37 UTC*, **@BertrandLeRoy** commented:

Did you ever try to reproduce this with a non-Azure instance? I've seen something similar and I'd like to be sure this is Azure specific.

**Comment by bleroy** [comment]
On *2011-08-05 09:45:29 UTC*, **@vrossi** commented:

Yes. We have a local development server that is set up with continuous integration. As builds are pushed to that environment and we update modules through the admin console, everything works great. We noticed this specifically in Azure because while we
 upgrading the same modules that were updated in the development environment, we would get SQL exceptions like, "Table already exists". That's when we would look at the schema in SQL Azure and notice that the tables were in fact created correctly
 but for whatever reason, the version was out of sync in the Orchard_Framework_DataMigrationRecord table.

So the module would always think it needed to be updated, even though it was. Just not according to it's version in the Orchard_Framework_DataMigrationRecord table.

**Comment by bleroy** [comment]
On *2011-08-05 09:48:07 UTC*, **@vrossi** commented:

Sorry, I meant to say that I am NOT able to reproduce this locally.

It works correctly locally. The issue seems to be Azure specific.

**Comment by bleroy** [comment]
On *2011-08-05 13:33:45 UTC*, **@BertrandLeRoy** commented:

Thanks.

**Comment by bleroy** [comment]
On *2011-08-09 20:54:59 UTC*, **@vrossi** commented:

So I stumbled across something while looking for some examples that might be some proof that others have come across this issue. If you look at the Migrations.cs in the Orchard.Blog module, you'll notice that they return 3 from all three updates, instead
 of returning 1, 2 and 3 respectively. That would certainly fix my issue, but it doesn't seem like the recommended process.

**Comment by bleroy** [comment]
On *2011-08-10 05:48:11 UTC*, **@sebastienros** commented:

Migration method can return any number, even future ones. The Blogs example you give here is normal. What happened is that it was initially with only Create() -> 1. Then we decided to add a new Part to the blogs, so we made a UpdateFrom1() -> 2.
 For existing users, it would run only this update method. But for new users it would run Create, then Update. It would have been useless as we knew that for new installations it would require the Part, so it was added to the Create method, and we then returned
 2, which is sort of a schema version. This way newly installed instances would retain they are at schema version 2, and updated instances are also at version 2. Same thing for a new version of the module, when we decided to change the default value. We had
 to take into account people coming from a fresh install, from version 1, and from version 2. So to mitigate the number of Update method calls, all existing ones have been updated and return directly to the version 3.

We could definitely have returned 1, 2 and 3 if we had not changed the content of those methods, it's just a matter of mood, the result is the same finally.
